### PR TITLE
fluent scale write

### PR DESCRIPTION
**Pull request by kalaninja**
adds fluent injections
adds fluent scale write

### CODE REVIEW

File: common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.strategyobject.substrateclient.common.inject;
```
Comment (vnabiev) [inline common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java:1]: I think a verb is not a good option as a package's name,

File: common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+public interface Inject<R, D> {
```
Comment (vnabiev) [inline common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java]: How about `Dependent`/`Dependable` or something like that?

File: common/src/main/java/com/strategyobject/substrateclient/common/inject/Injection.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+public class Injection<T, S extends Injection<T, S>> {
+    private final @NonNull T receiver;
+
+    private List<Injection<T, ?>> injections;
+
+    @SuppressWarnings("unchecked")
+    public S inject(@NonNull Object... injections) {
+        Preconditions.checkArgument(injections.length > 0);
+
+        this.injections = Arrays.stream(injections)
```
Comment (vnabiev) [inline common/src/main/java/com/strategyobject/substrateclient/common/inject/Injection.java:23]: I suppose the behaviour can be non-obvious as so it follows the principle of method chaining, but mutates the current state.

File: common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java (hunk 2 above)
Comment (vnabiev) [inline common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java]: I guess `R` should extend the interface.

File: scale/src/main/java/com/strategyobject/substrateclient/scale/ScaleUtils.java [hunk 4]
```diff
@@ -30,6 +32,17 @@ public static <T> byte[] toBytes(T value, @NonNull ScaleWriter<T> writer) {
         return stream.toByteArray();
     }
 
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public static <T> Injector<Class<?>, ScaleWriter, byte[]> toBytes(T value,
```
Comment (vnabiev) [inline scale/src/main/java/com/strategyobject/substrateclient/scale/ScaleUtils.java]: I would suggest to return `byte[]` from the method.

File: common/src/main/java/com/strategyobject/substrateclient/common/inject/Injection.java (hunk 3 above)
Comment (kalaninja) [reply common/src/main/java/com/strategyobject/substrateclient/common/inject/Injection.java:23]: It mimics the behavior of `ScaleWriter.inject()`

File: common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java (hunk 2 above)
Comment (kalaninja) [reply common/src/main/java/com/strategyobject/substrateclient/common/inject/Inject.java]: That would imply that the injection produces something that you can inject into again, which is not true. The opposite, I want to encourage to call inject only once. Currently, you can inject multiple `optionWriter.inject(...).inject(...).inject(...)` and according to the current implementation only the first injection is used, all the rest are ignored. So, I think it is not right to put it into the contract that R extends Inject.

File: scale/src/main/java/com/strategyobject/substrateclient/scale/ScaleUtils.java (hunk 4 above)
Comment (kalaninja) [reply scale/src/main/java/com/strategyobject/substrateclient/scale/ScaleUtils.java]: good idea

**Review by vnabiev — APPROVED** [review]

**Pull request merged**